### PR TITLE
[SW-1187] Use H2OMojoModel to load mojo & mojo pipelines as well

### PR DESCRIPTION
**Pull request by jakubhava**

### CODE REVIEW

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGridSearch.scala [hunk 1]
```diff
@@ -118,15 +118,16 @@ class H2OGridSearch(override val uid: String) extends Estimator[H2OMOJOModel]
...
     gridMojoModels = gridModels.map { m =>
       val data = ModelSerializationSupport.getMojoData(m)
-      new H2OMOJOModel(data, Identifiable.randomUID(s"${$(gridAlgoParams).algoName()}_mojoModel"))
+      H2OMOJOModel.createFromMojo(data, Identifiable.randomUID(s"${$(gridAlgoParams).algoName()}_mojoModel"))
     }
 
-    // Block until GridSearch finishes
-    val model = trainModel(grid)
+    val mojoData = ModelSerializationSupport.getMojoData(selectModelFromGrid(grid))
+    val model = H2OMOJOModel.createFromMojo(mojoData, Identifiable.randomUID("gridSearch_mojoModel"))
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGridSearch.scala]: The identifier should be specific to the concrete algorithm.

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOLoader.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    createFromMojo(is, Identifiable.randomUID(inputPath.getName))
+  }
+
+  def createFromMojo(is: InputStream, uid: String): T = {
+    createFromMojo(Stream.continually(is.read).takeWhile(_ != -1).map(_.toByte).toArray, uid)
+  }
+
+  def createFromMojo(mojoData: Array[Byte], uid: String): T
+}
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOLoader.scala]: nit: new line

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModel.scala [hunk 3]
```diff
@@ -225,85 +218,23 @@ class H2OMOJOModel(val mojoData: Array[Byte], override val uid: String)
...
     val mojoModel = ModelSerializationSupport.getMojoModel(mojoData)
-    val sparkMojoModel = m.runtimeClass.getConstructor(classOf[Array[Byte]], classOf[String]).
-      newInstance(mojoData, uid).asInstanceOf[T]
+    val model = new PyH2OMOJOModel(uid)
+    model.setMojoData(mojoData)
     // Reconstruct state of Spark H2O MOJO transformer based on H2O's Mojo
     if (mojoModel.isSupervised) {
-      sparkMojoModel.setFeaturesCols(mojoModel.getNames.filter(_ != mojoModel.getResponseName))
-      sparkMojoModel.setLabelCol(mojoModel.getResponseName)
+      model.setFeaturesCols(mojoModel.getNames.filter(_ != mojoModel.getResponseName))
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModel.scala]: What about filtering fold and weight column? Is it relevant?

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOReader.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+    val inputPath = new Path(path, serializedFileName)
+    val fs = inputPath.getFileSystem(SparkSession.builder().getOrCreate().sparkContext.hadoopConfiguration)
+    val qualifiedInputPath = inputPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
+    val is = fs.open(qualifiedInputPath)
+    val mojoData = Stream.continually(is.read()).takeWhile(_ != -1).map(_.toByte).toArray
+    model.setMojoData(mojoData)
+    model
+  }
+
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOReader.scala]: nit: remove new line

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOReader.scala [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+}
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOReader.scala]: nit: add new line

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOWriter.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.apache.spark.ml.h2o.models
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.ml.param.Params
+import org.apache.spark.ml.util.{DefaultParamsWriter, MLWriter}
+
+private[models] class H2OMOJOWriter(instance: Params, val mojoData: Array[Byte]) extends MLWriter {
+
+  val serializedFileName = "mojo_model"
```
Comment (mn-mikke) [inline ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOWriter.scala]: Would it be possible to specify ```"mojo_model"``` at one place? (loader vs. writer)

File: ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGridSearch.scala (hunk 1 above)
Comment (jakubhava) [reply ml/src/main/scala/org/apache/spark/ml/h2o/algos/H2OGridSearch.scala]: yup, thx!

File: ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModel.scala (hunk 3 above)
Comment (jakubhava) [reply ml/src/main/scala/org/apache/spark/ml/h2o/models/H2OMOJOModel.scala]: Good point -> we will solve this by using the MK change. Right now the values even though set, were not used anywhere

**Pull request merged**